mono/metadata/class-init.c

```diff
@@ -311,8 +311,13 @@ mono_class_setup_fields (MonoClass *klass)
 	 * Prevent infinite recursion by using a list in TLS.
 	 */
 	GSList *init_list = (GSList *)mono_native_tls_get_value (setup_fields_tls_id);
-	if (g_slist_find (init_list, klass))
+	if (g_slist_find (init_list, klass)) {
+		// Recursive value types will not resolve their size and assert later when we attempt to fetch it.
+		// Flag here to gracefully handle and trigger exception.
+		if (klass->valuetype && !klass->size_inited && !mono_is_corlib_image (klass->image))
+			mono_class_set_type_load_failure (klass, "Recursive type definition detected %s.%s", klass->name_space ? klass->name_space : "", klass->name);
```

Comment:
Should we use `mono_type_get_full_name` here to get the correct name for nested types?